File: TP/TP_python/etudiant.py
from personne import Personne

class Etudiant(Personne):

    def __init__(self, nom, age, numero, moyenne):
        super().__init__(nom, age)
        self.numero = numero
        self.list_cours = []
        self.moyenne = moyenne

    @property
    def moyenne(self):
        return self.__moyenne
    
    @moyenne.setter
    def moyenne(self, value):
        if not(0 <= value <= 20):
            raise ValueError("Moyenne invalide. Elle doit être entre 0 et 20.")
        self.__moyenne = value


    def __str__(self):
        return super().__str__() + f", Numero: {self.numero}, Moyenne: {self.moyenne}, Cours: {', '.join(str(cours) for cours in self.list_cours)}"
    

    def afficher_details(self):
        super().afficher_details()
        print(f"Moyenne: {self.moyenne}")

    def ajouter_cours(self, cours):
        self.list_cours.append(cours)
        

File: src/main/java/Etudiant.java
import java.util.List;
import java.util.ArrayList;

public class Etudiant extends Personne {
    private String numEtudiant;
    private double moyenne;
    private List<Cours> cours;

    public Etudiant(String nom, int age, String numEtudiant, double moyenne) {
        super(nom, age);
        this.numEtudiant = numEtudiant;
        this.setMoyenne(moyenne);
        this.cours = new ArrayList<>();
    }

    public void ajouterCours(Cours c) {
        this.cours.add(c);
    }

    public void setMoyenne(double m) {
        if (m < 0.0 || m > 20.0) {
            throw new IllegalArgumentException("La moyenne doit être entre 0 et 20");
        }
        this.moyenne = m;
    }

    @Override
    public String toString() {
        return super.toString() + " - Cours: " + cours;
    }
}